packages/flame_bloc/lib/src/flame_bloc_listenable.dart

```diff
@@ -0,0 +1,71 @@
+import 'dart:async';
+
+import 'package:bloc/bloc.dart';
+import 'package:flame/components.dart';
+import 'package:meta/meta.dart';
+
+import '../flame_bloc.dart';
+
+/// Adds [Bloc] access and listening to a [Component]
+mixin FlameBlocListenable<B extends BlocBase<S>, S> on Component {
+  late S _state;
+  late final StreamSubscription<S> _subscription;
+  B? _bloc;
+
+  /// Explicitly set the bloc instance which the component will be listening to.
+  /// This is useful in cases where the bloc being listened to
+  /// has not be provided via [FlameBlocProvider].
+  set bloc(B bloc) {
+    assert(
+      _bloc == null,
+      'Cannot update the bloc instance once it has been set.',
+    );
+    _bloc = bloc;
+  }
+
+  @override
+  @mustCallSuper
+  Future<void> onLoad() async {
+    var bloc = _bloc;
+    if (bloc == null) {
+      final providers = ancestors().whereType<FlameBlocProvider<B, S>>();
+      assert(
+        providers.isNotEmpty,
+        'No FlameBlocProvider<$B, $S> available on the component tree',
+      );
+
+      final provider = providers.first;
+      _bloc = bloc = provider.bloc;
+    }
+
+    _state = bloc.state;
+    _subscription = bloc.stream.listen((newState) {
+      if (_state != newState) {
+        final _callNewState = listenWhen(_state, newState);
+        _state = newState;
+
+        if (_callNewState) {
+          onNewState(newState);
+        }
+      }
+    });
+  }
+
+  /// Override this to make [onNewState] be called only when
+  /// a certain state change happens.
+  ///
+  /// Default implementation returns true.
+  bool listenWhen(S previousState, S newState) => true;
```

Comment:
Shouldn't we make this name more strict to a bloc context? When mixed into a component, this overrides of this may be lost among methods for other mixins. It would be more readable if it at least gave an idea that it is bloc-related.

Reply:
Any suggestion? `listenWhen` for me makes it pretty clear that it is related to a `Listenable` mixin, but I am open to suggestions

Reply:
In Flutter there's a method `didUpdateWidget`, so we could try something similar here: `didUpdateBloc` or `didUpdateBlocState`.

Reply:
IMO, `listenWhen` is better, `flutter_bloc` has a `listenWhen` attribute on their `BlocListener` so it will feel more in line for people used to flutter bloc

Reply:
I would suggest `filterStateUpdates` Based on:

- `listen`, in my interpretation implies that this map sets or unsets a configuration (for example: `addListener` or `removeListener`) from somewhere, when it is supposed to declare a logic that filter state updates from bloc (the component is technically always listening via `_subscription`).

- It aligns with `onNewState` about regarding `state`. 

Now the fact that `flutter_bloc` has a `listenWhen` invalidates all of this discussion given that the aparent reason is to keep it similar to that package, which is fine here (flame_bloc not flame).

Reply:
> Now the fact that flutter_bloc has a listenWhen invalidates all of this discussion given that the aparent reason is to keep it similar to that package, which is fine here (flame_bloc not flame).

Exactly, lets keep it as `listenWhen`